cli/xml-to-ts.ts

```diff
@@ -0,0 +1,236 @@
+${padding}}>`
+      : "";
+  return `${padding}${f.name}${optional}: ${f.type}${subfields}${comment}`;
+}
+
+/**
+ * xmlToIface parses an xml-file and returns an iface
+ * @param name the name of the interface
+ * @param xml
+ */
+function xmlToIface(name: string, xml: string): iface {
+  const iface: iface = {
+    name,
+    fields: []
+  };
+  const doc = new xmldom.DOMParser().parseFromString(xml);
+  const form = xpath
+    .evaluate("//form", doc, null, xpath.XPathResult.ANY_TYPE, null)
+    .iterateNext();
+  iface.fields.push(...getInputFields(form));
+  iface.fields.push(...getFieldSetItems(form));
+  iface.fields.push(...getItemSetFields(form));
+  return iface;
+}
+
+function getInputFields(node): field[] {
+  const fields = [];
+  const inputs = xpath.evaluate(
+    "./input",
+    node,
+    null,
+    xpath.XPathResult.ANY_TYPE,
+    null
+  );
+  let input: XPathResult = null;
+  while ((input = inputs.iterateNext())) {
+    fields.push(createFieldFromInput(input));
+  }
+  return fields;
+}
+
+// FieldSet is a flat structure with no nesting.
+function getFieldSetItems(node): field[] {
+  const fields = [];
+  const fieldSets = xpath.evaluate(
+    "./field-set",
+    node,
+    null,
+    xpath.XPathResult.ANY_TYPE,
+    null
+  );
+
+  let fieldSet = null;
+  while ((fieldSet = fieldSets.iterateNext())) {
+    const items = xpath
+      .evaluate("./items", fieldSet, null, xpath.XPathResult.ANY_TYPE, null)
+      .iterateNext();
+    const inputs = xpath.evaluate(
+      "./input",
+      items,
+      null,
+      xpath.XPathResult.ANY_TYPE,
+      null
+    );
+
+    let input = null;
+    while ((input = inputs.iterateNext())) {
+      fields.push(createFieldFromInput(input));
+    }
+  }
+  return fields;
+}
+
+// ItemSet is a nested structure.
+function getItemSetFields(node): field[] {
+  const fields = [];
+  const itemSets = xpath.evaluate(
+    "./item-set",
+    node,
+    null,
+    xpath.XPathResult.ANY_TYPE,
+    null
+  );
+
+  let itemSet = null;
+  while ((itemSet = itemSets.iterateNext())) {
+    const name = xpath.select1("@name", itemSet).value;
+    const optional =
+      xpath.select1("./occurrences/@minimum", itemSet).value == 0;
+
+    const items = xpath
+      .evaluate("./items", itemSet, null, xpath.XPathResult.ANY_TYPE, null)
+      .iterateNext();
+    const subfields = [
+      ...getInputFields(items),
+      ...getFieldSetItems(items),
+      ...getItemSetFields(items)
+    ];
+
+    fields.push({ name, type: "Array", optional, subfields });
+  }
+  return fields;
+}
+
+/**
+ * Create a field from an <input> node.
+ * @param input an xpath input node
+ */
+function createFieldFromInput(input): field {
+  const name = xpath.select1("@name", input).value;
+  const comment = xpath.select1("string(./label)", input);
+  const optional = xpath.select1("./occurrences/@minimum", input).value == 0;
+  const type = "string";
+  return { name, type, comment, optional };
+}
+
```

Comment:
I think this file is a bit large, and should be split into multiple files around here. Maybe pull the code above out, and make  it a tool used by the cli, and IO stuff underneath.